### PR TITLE
[4.0] Remove JError

### PR DESCRIPTION
**Pull request by laoneo**
JError is deprecated since ages and it was marked to be removed in Joomla 4 in the deprecate message. This pr cleans up the core from JError and removes the class.
It is the followup pr of #16952 and #16967 which is the last piece to have `JException`and `JError` being removed.

### CODE REVIEW

File: libraries/fof/platform/platform.php [hunk 1]
```diff
@@ -354,10 +354,6 @@ public function setIntegrationObject($key, $object)
 	 */
 	public function setErrorHandling($level, $log_level, $options = array())
 	{
-		if (version_compare(JVERSION, '3.0', 'lt') )
```
Comment (C-Lodder) [inline libraries/fof/platform/platform.php]: Don't need to edit FOF do we?

Comment (brianteeman) [reply libraries/fof/platform/platform.php]: Definitely not. It's a vendor file.

Comment (laoneo) [reply libraries/fof/platform/platform.php]: As far as I know, Joomla is maintaining the FOF code till it gets removed. So we should keep it clean to a degree where it is not breaking obviously.

**Comment by wilsonge** [comment]
can you fix conflicts here please?

File: libraries/fof/platform/platform.php (hunk 1 above)
Comment (wilsonge) [reply libraries/fof/platform/platform.php]: It's being removed https://github.com/joomla/joomla-cms/pull/17687 :P so not sure it matters for changes in 4.x

**Comment by laoneo** [comment]
Conflicts fixed

**Comment by laoneo** [comment]
Conflicts fixed

**Comment by laoneo** [comment]
What is the status of this one?

**Comment by mbabker** [comment]
Are we positive that the use of `getError` and `setError` in JObject subclasses (or any place doing this still really) are OK with proper Exceptions and have no internal couplings to JError or JException mechanisms?  If we can drop this side of the functionality without completely breaking those two methods which still have heavy use I think it's worth re-syncing this PR and doing a heavy testing round to make sure we're OK.

**Comment by laoneo** [comment]
I'm going to resync when I get the ok from @wilsonge that it is getting in when all is fine.

**Comment by wilsonge** [comment]
`JObject`'s `setError` method definitely just adds (what's doc-block documneted as) strings from an internal array  https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Object/CMSObject.php#L229-L243 . In the getters we also deal with that string actually being an exception https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Object/CMSObject.php#L131-L182 

So we're definitely safe for JObject classes. I guess I need to go back through this again and just check what classes are actually using JError directly - but most of this stuff is outdated docblocks iirc

**Comment by laoneo** [comment]
As I did this pr, I removed every trace of JError.

**Comment by laoneo** [comment]
Was a rather painless merge, so we are in sync again.

**Comment by wilsonge** [comment]
On review this looks fine to me. Let's get some heavy testing on this

**Comment by wilsonge** [comment]
OK Well I'm sure we'll break some things - but clearly we need this and it's not getting tests

**Comment by skurvish** [comment]
In my component migration to namespaces in preparation of J4 I am finding JError in src/Mail/Mail.php so it looks like not all traces of JError are gone.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17669">issues.joomla.org/tracker/joomla-cms/17669</a>.</sub>

**Comment by laoneo** [comment]
@skurvish can you post here a link to the line you see this error? I can't find a reference to the class, just an entry of the logger with jerror.

**Comment by skurvish** [comment]
In the current joomla git you can find JError at lines 118 & 154 in Mail/Mail.ph. You can also find it in Application/CMSApplication.php, Cache/Storage/RedisStorage.php, Client/ClientWrapper.php, Installer/Adapter/ComponentAdapter.php, Menu/SiteMenu.php, MVC/View/CategoryFeedView.php, MVC/Model/AdminModel.php & MVC/View/CategoryView.php. These are all in libraries/src

**Comment by mbabker** [comment]
Are you looking at the right version of those files?  It’s expected to still be there for 3.x versions, but should be gone in 4.0 (just looking at the Mail.php file it isn’t used in the 4.0 file).

**Comment by skurvish** [comment]
Current github and my installed 3.9 version. Problem I am having is that I am running a namespaced WebApplication which I have loaded the framework (included defines.php & framework.php), created a class of WebApplication. If an error happens in Mail.php it is calling \JError which is not found. 

Perhaps there is some other autoloader I need to call beforehand so that JError is found?

**Comment by wilsonge** [comment]
In J3 it is expected JError exists. We're only removing it in J4. You should be covered by the framework.php include which in turn loads the autoloaders in `libraries/import.legacy.php` which should autoload everything in libraries/legacy

**Comment by skurvish** [comment]
Hmmm, but it doesn't seem to autoload JError. Where should I look for answers? I assume this area is solely for J4.

**Comment by wilsonge** [comment]
This Pull Request was to the J4 branch (hence why we assumed you were referring to J4 in your comment here). If you create a fresh issue and drop in a stack trace of the issue your getting we should be able to help more :)

**Comment by skurvish** [comment]
OK. I fixed the reason JError was being called so will have to revert to the state which caused the error and then will post a fresh issue. Thanks.

**Pull request merged**